File: BOJ(Java 공부용)/15666.java
import java.io.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int m = Integer.parseInt(input[1]);
        String[] input2 = br.readLine().split(" ");
        ArrayList<Integer> v = new ArrayList<>();
        ArrayList<Integer> temp = new ArrayList<>();
        HashSet<ArrayList<Integer>> set = new HashSet<>();
        for(int i=0;i<input2.length;i++) {
            v.add(Integer.parseInt(input2[i]));
        }
        Collections.sort(v);
        go(0,0,m,v,temp,set);
        bw.flush();
        bw.close();
    }
    private static void go(int here,int cnt,int depth,ArrayList<Integer> v,ArrayList<Integer> temp,HashSet<ArrayList<Integer>> set) throws IOException {
        if(cnt == depth) {
            if(!set.contains(temp)) {
                set.add(temp);
                for(int i=0;i<temp.size();i++) {
                    bw.write(temp.get(i)+" ");
                }
                bw.write("\n");
            }
            return;
        }
        for(int i=here;i<v.size();i++) {
                temp.add(v.get(i));
                go(i,cnt + 1, depth, v, temp, set);
                temp.remove(temp.size() - 1);
        }
    }
}

File: BOJ(Kotlin)/15666.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val bw = BufferedWriter(OutputStreamWriter(System.out))
fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()
    st = StringTokenizer(br.readLine())
    val array = mutableListOf<Int>()
    for (i in 0 until n) {
        array.add(st.nextToken().toInt())
    }
    array.sort()
    val permutationArray = mutableListOf<Int>()
    dfs(0,n, m, 0, array, permutationArray)
    bw.flush()
    bw.close()

}

fun dfs(index : Int,
    n: Int,
    m: Int,
    count: Int,
    array: MutableList<Int>,
    permutationArray: MutableList<Int>,
) {
    if (count == m) {
        for (value in permutationArray) {
            bw.write("$value ")
        }
        bw.write("\n")
        return
    }

    var prevNum = -1
    for (i in index until n) {
        if (prevNum != array[i]) {
            prevNum = array[i]
            permutationArray.add(array[i])
            dfs(i,n, m, count + 1, array, permutationArray)
            permutationArray.removeLast()
        }
    }

}